4.0 beta article (#886)
withastro/astro.build: src/content/blog/astro-4-beta.mdx

---
title: "Astro 4.0 Beta"
description: "Astro 4.0 beta is released."
publishDate: "November 22, 2023"
authors:
  - matthew
coverImage: "/src/content/blog/_images/astro-350/post-header-3.5.webp"
socialImage: "/src/content/blog/_images/astro-350/og-image-3.5.webp"
lang: "en"
---

> Today we've released the first beta of Astro 4! Those feeling adventurous can install new beta, review our in-progress migration docs, and report any bugs you find to the [Astro discord](https://astro.build/chat).

Reviewer: ```suggestion
Today we've released the first beta of Astro 4! Those feeling adventurous can install the new beta, review our in-progress upgrade docs, and report any bugs you find to the [Astro discord](https://astro.build/chat).
```

## What to expect in Astro 4

> We're releasing Astro 4 to stay up to date with the latest version of Vite, which just released Vite 5 last month.

Reviewer 2: ```suggestion
We are releasing Astro 4 to stay up to date with [Vite](https://vitejs.dev/), our internal build tool. Astro 4 includes the latest version of [Vite 5](https://vitejs.dev/blog/announcing-vite5), which was just released last month.
```

You could maybe trim this down a bit more from my suggestion! The main suggestion here is to quickly mention what Vite is and what this means to the Astro user/reader (aka, "so what?").

> As Vite 5 is *also* a small major, most projects will be able to upgrade to Astro 4 with only minimal changes to your code, if any at all.

Reviewer 3: ```suggestion
As Vite 5 is *also* a small major, most projects will be able to upgrade to Astro 4 with only minimal changes to your code if any at all.
```

> Additionally with Astro 4 we are planning to bring some of the features we have been working on over the last few months out of experimental status and mark them as stable. We are still deciding *which* features those will be. Stay tuned for more!

Reviewer: ```suggestion
Additionally, in Astro 4 we are planning to bring some of the new features we have been working on over the last few months out of experimental status and mark them as stable. We are still deciding *which* features those will be. Stay tuned for more!
```

> Additionally, in Astro 4 we are planning to bring some of the new features we have been working on over the last few months out of experimental status and mark them as stable. We are still deciding *which* features those will be. Stay tuned for more!

Reviewer 2: ```suggestion
Additionally, in Astro 4 we are planning to bring some of the new features we have been working on over the last few months out of experimental status and mark them as stable. Stay tuned for more!
```

It's a funny line, but it also reads a bit like "ehh, we don't know what we're doing!" Would suggest removing

## Trying out the beta

You can try out the new beta in your existing app by installing the `astro@beta` package with your favorite package manager.

```sh
npm install astro@beta
pnpm install astro@beta
yarn install astro@beta
```

If starting a new project you can also try out the beta by running:

```sh
npm init astro@beta
```

> Again, substitute with your favorite package manager. We have not completed the migration guide yet, but you can see what is being changed in the [pull request] (LINK NEEDED)

Reviewer: ```suggestion
Again, substitute with your favorite package manager. We have not completed the upgrade guide yet, but you can already find the most significant changes in the [Astro 4.0 upgrade guide pull request](https://github.com/withastro/docs/pull/5481). You can also follow its related issue which is tracking all breaking changes.
```
Maybe something like this?

## Astro 3.6

> Additionally we've recently released Astro 3.6 which comes with new events in the `<ViewTransitions />` router as well as fallback for `prefetch` in reduced data scenarios.

Reviewer 3: ```suggestion
Additionally, we've recently released Astro 3.6 which comes with new events in the `<ViewTransitions />` router as well as fallback for `prefetch` in reduced data scenarios.
```

### View Transition events

Astro 3.6 adds these three new events as part of the `<ViewTransitions />` router's lifecycle:

> `astro:before-preparation` runs immediately before the *preparation* step, where the content of the new page is loaded. This can be used to override the behavior of loading the new page, for example to show a loading spinner.

Reviewer 3: ```suggestion
`astro:before-preparation` runs immediately before the *preparation* step, where the content of the new page is loaded. This can be used to override the behavior of loading the new page, for example, to show a loading spinner.
```

> `astro:before-preparation` runs immediately before the *preparation* step; where the content of the new page is loaded. This can be used to override the behavior of loading the new page, for example to show a loading spinner.
> 
> `astro:after-preparation` runs after the content has been loaded. This can be used to *stop* a loading spinner, if one was added as part of `astro:before-preparation`.
> 
> `astro:before-swap` runs immediately before the *swap* step; where the contents of the old page are swapped out in favor of the contents of the new page. This event includes the `event.newDocument` property, a `Document` that you can mutate. One use-case for this event is to set dark mode class names if using `localStorage` to store light/dark mode preferences.

Reviewer: ```suggestion
`astro:before-preparation` runs immediately before the *preparation* step, where the content of the new page is loaded. This can be used to override the behavior of loading the new page, for example to show a loading spinner.

`astro:after-preparation` runs after the content has been loaded. This can be used to *stop* a loading spinner if one was added as part of `astro:before-preparation`.

`astro:before-swap` runs immediately before the *swap* step, where the contents of the old page are swapped out in favor of the contents of the new page. This event includes the `event.newDocument` property, a `Document` that you can mutate. One use-case for this event is to set dark mode class names if using `localStorage` to store light/dark mode preferences.
```
grammar nits only!

> Check out [this demo](https://events-3bg.pages.dev/) which shows what is possible using the new events.

Reviewer: ```suggestion
Check out [this demo](https://events-3bg.pages.dev/) which shows what is possible using the new events and see the docs for more information about [all five lifecycle events](https://docs.astro.build/en/guides/view-transitions/#lifecycle-events).
```
I think some kind of line and link to the entire lifecycle makes sense, since we're showing only three of them here, and this really is part of a larger poicture.

### Prefetch fallback

> Previously when a user was on a slow connection or had the data saver preference enabled there would be no prefetch. In 3.6 we've changed it so that instead prefetch occurs on `tap`. A `tap` occurs when the user begins clicking a link, but before they have released. This gives us just a little bit of a head start on fetching the next page, but only when we're pretty sure the user intends to click.

Reviewer: ```suggestion
Previously when a user was on a slow connection or had the data saver preference enabled, prefetching was entirely disabled. In 3.6 we've changed the prefetch strategy to occur on `tap`: when the user begins clicking a link, but before they have released. This gives us just a little bit of a head start on fetching the next page, but only when we're pretty sure the user intends to click.
```
